### PR TITLE
Core devel

### PR DESCRIPTION
**Pull request by jjgomezcadenas**
A major revision of the core modules has been conducted. This includes cleaning up, eliminating many obsolete functions, as well as re-organizing functions, across modules. 

core_functions.py	-- general purpose functions. It has been cleaned up, and a simple test suite has been added. There is also a doc notebook (which is not included in this commit, waiting for policy on notebooks and docs to be fully agreed). 

core_functions_performance.py: a module with a simple example case of performance of python vs Cython for a particular example.

mpl_functions.py --- general purpose plotting functions. No tests are relevant here, but the plotting functions are exercised in the documentation notebooks (diomira, irene) which are ready to be committed.

system_of_units.py -- definition of system of units. Extensively tried and tested, stable

random_sampling.py -- sampler of SiPM pdf functions. Diomira-notebook shows that it behaves as expected but a test would be nice (Gonzalo Martinez, pending)

log_config.py --- configures the logger. Not used currently, possible candidate for deletion.

configure.py -- utility used for configuration, works as expected, stable. 

sensor_functions.py -- collects functions that manipulate sensors. Most of them exercised in notebooks. A test could be added

tbl_functions.py -- utility for pYtables manipulation

wfm_functions.py --- waveform manipulation. There is a simple test suite (can be enlarged)

wfm_to_df_functions.py -- collection of functions where wmf are transformed to data frames for manipulation. Not used at the moment. Possible candidate for deletion. 

mctrk_functions.py -- for manipulation of MC "true" tracks. Not being used at the moment (reconstruction does not use "true" variables) but eventually useful.

peak_functions.py. --- to be revised in next cycle, associated to Irene
pmaps_functions.py --- to be revised in next cycle, associated to Irene

### CODE REVIEW

**Comment by jacg** [comment]
Please ensure that commits you submit for review form logical units.

The "work in progress" commit shouldn't be there: you only put it there because you couldn't be bothered to stash when changing branches. It's rude to ask the reviewer to look at this kind of mess.

I see that data files addressing the #81 Irene bug have been added to the repo, but I can't find any commit message referring to it at all. Please put the bugfix and its tests in a single commit with no other stuff in it. Alternatively put the tests in one commit, and the fix in another. Without any other stuff polluting those commits.

When that's done, I'll have a look at the code itself.

**Comment by jacg** [comment]
It's rather suspicious that, before including these commits, pytest finds 14 tests, and after including these commits (which include a new test for the irere bug) pytest still finds only 14 tests.

**Comment by jjgomezcadenas** [comment]
There is no single fix to core, but quite a few, reflecting several different actions over the core module. It is true that the first commit is rush (because we were in a rush and I was advised to make a quick commit and do something else) and that the files for the bug fix are there, but the rests of the commits make sense, I think. So please, kindly explain me what exactly I am supposed to do. Step by step, if possible.

**Comment by jjgomezcadenas** [comment]
Incidentally, the pR included a very extensive message to explain what was the status of the core module after the changes and make easier the reviewer to understand the work done. Plus a new test module.  Perhaps the commits were a bit disorganised (in the sense that the first commit was a bit off hand and the files for the Irene test sneaked it) but I wouldn’t call it a mess.

**Comment by jacg** [comment]
> So please, kindly explain me what exactly I am supposed to do. Step by step, if possible.

This is best done live.

> the pR included a very extensive message

Yes, the PR message was good. The commit messages could be much better. And so could the diffs, probably.

>  It is true that the first commit is rush (because we were in a rush and I was advised to make a quick commit and do something else)

You will recall that the advice was "You can commit with 'work in progress' as the commit message, because this is in your local repo and you will clean it up before submitting a pull request, so nobody in the outside world will ever see the 'work in progress' commit message. And yet, I can see 'Work in progress' as a commit message in a PR submitted to the certral repo. That is the *opposite* of the advice you were given!

> Perhaps the commits were a bit disorganised [...] but I wouldn’t call it a mess.

For me, that's self-contradictory.

But, fundamentally, you have to think about how seriously we want to take the reviewing process. If the commits are difficult to follow (unhelpful messsages, diffs which do not logically partition or group together the work done) then the reviewer has two major options:

- Just approve it without understanding what is going on.
- Do something less frustrating with the precious time available and hope that someone else reviews it, or hope that he might find the time and will to do it at some time in the future.

In the first case, we might aswell not have a review process. In the second case PRs will take too long to get approved and take too much of the reviewer's precious time.

In short, please take the time to *learn how to* prepare good PRs and then take the time to *make* good PRs. I believe that, in the long run, this will benefit everyone taking part in the project. Alternatively, let's just forget the review process.

Bonus: making the effort to prepare good PRs often makes you notice and learn things about the code/work, that you wouldn't notice otherwise. This is often valuable.

I understand the frustration we experience when we are in a hurry, and the effort needed to make a good PR seems like an pointless obstacle. But bad PRs do accumulate technical debt, and you will have to repay it sooner or later. With interest. In the long run it is usually cheaper not to accumulate tecnical debt in the first place.

**Comment by jacg** [comment]
Incidentally, in relation to the above, there is a *lot* to be said for pair programming. Perhaps we should consider imposing it as the *only* way to work on the project, exept for those who have earned their *IC Solo Programming Licence* :-)

**Comment by jjgomezcadenas** [comment]
Not sure what part programming is, but if it implies teams of two working in the same project, I am afraid we are way too short of resources…

Also I am not sure the problem here is programming but deficient understanding of a difficult tool (git).

**Review by jacg — APPROVED** [review]
